src/unitquaternion.jl

```diff
@@ -0,0 +1,428 @@
+inv(q::UnitQuaternion) = conj(q)
+(-)(q::UnitQuaternion{T,D}) where {T,D} = UnitQuaternion{T,D}(-q.w, -q.x, -q.y, -q.z)
+
+# Norms
+LinearAlgebra.norm(q::UnitQuaternion) = sqrt(q.w^2 + q.x^2 + q.y^2 + q.z^2)
+vecnorm(q::UnitQuaternion) = sqrt(q.x^2 + q.y^2 + q.z^2)
+
+function LinearAlgebra.normalize(q::UnitQuaternion{T,D}) where {T,D}
+    n = 1/norm(q)
+    UnitQuaternion{T,D}(q.w*n, q.x*n, q.y*n, q.z*n)
+end
+
+# Identity
+(::Type{Q})(I::UniformScaling) where Q <: UnitQuaternion = one(Q)
+
+# # Equality
+# (≈)(q::UnitQuaternion, u::UnitQuaternion) = q.w ≈ u.w && q.x ≈ u.x && q.y ≈ u.y && q.z ≈ u.z
+# (==)(q::UnitQuaternion, u::UnitQuaternion) = q.w == u.w && q.x == u.x && q.y == u.y && q.z == u.z
+
+# Exponentials and Logarithms
+function exp(q::UnitQuaternion{T,D}) where {T,D}
+    θ = vecnorm(q)
+    sθ,cθ = sincos(θ)
+    es = exp(q.w)
+    M = es*sθ/θ
+    UnitQuaternion{T,D}(es*cθ, q.x*M, q.y*M, q.z*M, false)
+end
+
+function expm(ϕ::SVector{3,T}) where T
+    θ = norm(ϕ)
+    sθ,cθ = sincos(θ/2)
+    M = 0.5*sinc(θ/2π)
+    UnitQuaternion{T,ExponentialMap}(cθ, ϕ[1]*M, ϕ[2]*M, ϕ[3]*M, false)
+end
+
+function log(q::UnitQuaternion{T,D}, eps=1e-6) where {T,D}
+    # Assumes unit quaternion
+    θ = vecnorm(q)
+    if θ > eps
+        M = atan(θ, q.w)/θ
+    else
+        M = (1-(θ^2/(3q.w^2)))/q.w
+    end
+    UnitQuaternion{T,D}(0.0, q.x*M, q.y*M, q.z*M, false)
+end
+
+function logm(q::UnitQuaternion{T}) where T
+    # Assumes unit quaternion
+    q = log(q)
+    SVector{3,T}(2*q.x, 2*q.y, 2*q.z)
+end
+
+# Composition
+""" Quternion Composition
+Equivalent to
+```julia
+Lmult(q) * SVector(w)
+```julia
+Rmult(w) * SVector(q)
+```
+"""
+function (*)(q::UnitQuaternion{T1,D1}, w::UnitQuaternion{T2,D2}) where {T1,T2,D1,D2}
+    T = promote_type(T1, T2)
+    D = D2
+    UnitQuaternion{T,D}(q.w * w.w - q.x * w.x - q.y * w.y - q.z * w.z,
+                        q.w * w.x + q.x * w.w + q.y * w.z - q.z * w.y,
+                        q.w * w.y - q.x * w.z + q.y * w.w + q.z * w.x,
+                        q.w * w.z + q.x * w.y - q.y * w.x + q.z * w.w, false)
+end
+
+""" Rotate a vector
+Equivalent to `Hmat()' Lmult(q) * Rmult(q)' Hmat() * r`
+"""
+function Base.:*(q::UnitQuaternion{Tq}, r::SVector{3}) where Tq
+    w = q.w
+    v = vector(q)
+    (w^2 - v'v)*r + 2*v*(v'r) + 2*w*cross(v,r)
+end
+
+"Scalar multiplication"
+function (*)(q::Q, w::Real) where Q<:UnitQuaternion
+    return Q(q.w*w, q.x*w, q.y*w, q.z*w, false)
+end
+(*)(w::Real, q::Q) where Q<:UnitQuaternion = q*w
+
+
+
+"Inverted composition. Equivalent to inv(q1)*q2"
+(\)(q1::UnitQuaternion, q2::UnitQuaternion) = conj(q1)*q2
+
+"Inverted composition. Equivalent to q1*inv(q2)"
+(/)(q1::UnitQuaternion, q2::UnitQuaternion) = q1*conj(q2)
+
+"Inverted rotation. Equivalent to inv(q)*r"
+(\)(q::UnitQuaternion, r::SVector{3}) = conj(q)*r
+
+
+# ~~~~~~~~~~~~~~~ Quaternion Differences ~~~~~~~~~~~~~~~ #
+function (⊖)(q::UnitQuaternion{T,D}, q0::UnitQuaternion) where {T,D}
+    D(q0\q)
```

Comment:
Yeah, this is a great point. This was an early design decision that, looking back, wasn't a great one. I've now removed D as a type parameter and made a separate `RotationError{T,D} <: StaticVector{3,T}`.

I think the new version is much cleaner and more general. It allows any rotation representation to be converted to a rotation error via any of the maps, just by converting to `UnitQuaternion` first. Plus this formulation makes more sense since the map is more naturally tied to the rotation error (since it defines how to convert back to a rotation), and not to a particular rotation.